{DX-1499} Gateway 5.5 release notes (#5043)
TykTechnologies/tyk-docs: tyk-docs/content/product-stack/tyk-gateway/release-notes/version-5.5.md

### Deprecations
<!-- Required. Use the following statement if there are no deprecations, or explain if there are -->
There are no deprecations in this release.

<!-- Optional section!
Used to share and notify users about our plan to deprecate features, configs etc. 
Once you put an item in this section, we must keep this item listed in all the following releases till the deprecation happens. -->
<!-- ##### Future deprecations
-->

### Upgrade instructions {#upgrade-5.5.0}
If you are upgrading to 5.5.0, please follow the detailed [upgrade instructions](#upgrading-tyk).

### Release Highlights
<!-- Required. Use similar ToV to previous release notes. For example for a patch release:
This release primarily focuses on bug fixes.
For a comprehensive list of changes, please refer to the detailed [changelog]({{< ref "#Changelog-vX.Y.0">}}) below.
-->
We are thrilled to introduce Tyk Gateway 5.5, bringing advanced rate-limiting capabilities, enhanced certificate authentication, and performance optimizations. For a comprehensive list of changes, please refer to the [changelog](#Changelog-v5.5.0) below.

> ### Per Endpoint Rate Limiting

Reviewer: ```suggestion
#### Per Endpoint Rate Limiting
```

> Now configure rate limits at the endpoint level for both Tyk OAS and Tyk Classic APIs, providing granular protection for upstream services against abuse.

Reviewer 2: Reworded and included links in this release highlights paragraph

> ### Root CA Support for Client Certificates

Reviewer: ```suggestion
#### Root CA Support for Client Certificates
```

> Simplify certificate management with support for root Certificate Authority (CA) certificates, enabling client authentication via certificates signed by the configured root CA.

Reviewer 2: updated wording and included links

> ### Optimised AST Document Handling

Reviewer: ```suggestion
#### Optimised AST Document Handling
```

Here it is important to explain the benefit of each changelog item. As mentioned by James in a previous Slack message (https://tyktech.slack.com/archives/C044R3ZTN6L/p1686812207060839?thread_ts=1686762128.651249&cid=C044R3ZTN6L):
"...it is important to document the customer impact for the work delivered, so we can share it with prospects/install base. For example:
"New Chart delivers x and y benefit to a and b customer use cases. The business impact for them will be this and that" -->

#### Added
<!-- This section should be a bullet point list of new features. Explain:

- The purpose of the new feature
- How does the new feature benefit users?
- Link to documentation of the new feature
- For OSS - Link to the corresponding issue if possible on GitHub to allow the users to see further info.

Each change log item should be expandable. The first line summarises the changelog entry. It should be then possible to expand this to reveal further details about the changelog item. This is achieved using HTML as shown in the example below. -->
<ul>
<li>
<details>
<summary>Added root CA support for client certificate authentication</summary>

> We've added support for using root Certificate Authority (CA) certificates as client certificates in API definitions. The system now authenticates clients with certificates signed by the configured root CA, simplifying certificate management for multiple clients sharing a common CA. This functionality is available for static mutual TLS (mTLS) configurations only.

Reviewer 2: Updated wording in the HTML section but the links display as raw markdown links. Have incorporated suggested wording and added links to release highlights paragraph.

</details>
</li>
<li>
<details>
<summary>Optimised creation and usage of AST documents in GQL library</summary>

Optimised the creation and usage of AST documents in our GQL library to reduce significant memory allocations caused by pre-allocations during initial creation. These optimizations free up resources more efficiently, minimising performance penalties with increased requests to the Gateway.
</details>
</li>
<li>
<details>
<summary>Implemented upstream Endpoint rate limits</summary>
 

> Introduced new controls to configure rate limits at the endpoint level for both Tyk OAS and Tyk Classic APIs. This feature allows for more granular protection of upstream services from abuse, similar to Tyk’s traditional 'API-level' or 'Global' rate limits but applied specifically at the endpoint level.

Reviewer 2: Updated wording in the change log item but discovered that the links are rendered within the HTML block as raw markdown so have removed the links. I have added the links to the release highlights block for endpoint request rate limiting

> <summary>Implemented upstream Endpoint rate limits</summary>

Reviewer: ```suggestion
<summary>Implemented upstream endpoint rate limits</summary>
```

</details>
</li>
<li>
<details>

> <summary>Enhanced API versioning with fallback option for <i>First of Path</i> version data location</summary>

Reviewer: ```suggestion
<summary>Improved handling of requests to non-existent versions of APIs when using URL path versioning</summary>
```

> Introduced a new configuration for API versioning with `url_versioning_pattern`. This regex-based field helps identify if the first part of a URL is a version pattern, improving control over `strip_versioning_data` and `fallback_to_default` functionalities.

Reviewer: ```suggestion
When using the URL path to indicate the API version (for example `/v1/my-api`) it is common to strip the version identifier (e.g. `/v1`) from the path before proxying the request to the upstream. If the client doesn't provide any version identifier this could lead to an invalid target URL and failed requests, rather than correctly redirecting to the default version. We have introduced an optional configuration `url_versioning_pattern` where you can specify a regex that Tyk will use to identify if the URL contains a version identifier and avoiding the accidental stripping of valid upstream path. For more detail, see [here]({{< ref "product-stack/tyk-gateway/advanced-configurations/api-versioning/api-versioning#stripping-url-path-version-and-default-fallback" >}}).
```

Reviewer 2: Updated with suggestion but had to remove links since the raw markdown link is displayed in the HTML.

</details>
</li>
</ul>

#### Fixed
<!-- This section should be a bullet point list that describes the issues fixed in the release. For each fixed issue explain:

- What problem the issue caused
- How was the issue fixed
- Link to (new) documentation created as a result of a fix. For example, a new configuration parameter may have been introduced and documented for the fix
- For OSS - Link to the corresponding issue if possible on GitHub to allow the users to see further info.

Each change log item should be expandable. The first line summarises the changelog entry. It should be then possible to expand this to reveal further details about the changelog item. This is achieved using HTML as shown in the example below. -->
<ul>
<li>
<details>

> <summary>Corrected ordering of OAS API paths to prevent middleware misapplication</summary>

Reviewer: ```suggestion
<summary>Fixed an issue where transformation middleware could incorrectly be applied to Tyk OAS API endpoints with nested paths</summary>
```

> Fixed an issue where nested API endpoints, such as '/test' and '/test/abc', might incorrectly apply middleware from the parent path to the nested path. The fix ensures that API endpoint definitions are correctly ordered, preventing this middleware misapplication and ensuring both the HTTP method and URL match accurately.

Reviewer: ```suggestion
Fixed an issue when using Tyk OAS APIs where nested API endpoints, such as '/test' and '/test/abc', might incorrectly apply middleware from the parent path to the nested path. The fix ensures that API endpoint definitions are correctly ordered so that the standard behaviour of Tyk is followed, whereby path matching is performed starting from the longest path, preventing middleware misapplication and ensuring both the HTTP method and URL match accurately.
```

</details>
</li>
<li>
<details>

> <summary>Optimised key creation process to reduce Redis DeleteRawKey commands</summary>

Reviewer: ```suggestion
<summary>Optimised key creation process to avoid unnecessary Redis `DeleteRawKey` commands</summary>
```

> Previously, key creation or reset led to an exponential number of Redis DeleteRawKey commands, especially problematic for access lists with over 100 entries. The key creation sequence now runs only once, eliminating redundant deletion of non-existent keys in Redis. This optimization significantly reduces deletion events, enhancing performance and stability for larger access lists.

Reviewer: ```suggestion
Previously, key creation or reset led to an exponential number of Redis `DeleteRawKey` commands,; this was especially problematic for access lists with over 100 entries. The key creation sequence now runs only once, eliminating redundant deletion of non-existent keys in Redis. This optimization significantly reduces deletion events, enhancing performance and stability for larger access lists.
```

</details>
</li>
<li>
<details>
<summary>Resolved SSE Streaming issue</summary>

Addressed a bug that caused Server Side Event (SSE) streaming responses to be considered for caching, which required buffering the response and prevented SSE from being correctly proxied.
</details>
</li>
<li>

>   <details>

Reviewer: ```suggestion
<details>
```

>   <summary>Fixed Analytics Latency Reporting for MDCB Setups</summary>

Reviewer: ```suggestion
<summary>Fixed analytics latency reporting for MDCB setups</summary>
```

> <summary>Resolved SSE Streaming issue</summary>

Reviewer: ```suggestion
<summary>Resolved SSE streaming issue</summary>
```

>  Resolved an issue where Host and Latency fields (Total and Upstream) were not correctly reported for edge gateways in MDCB setups. The fix ensures accurate Host values and Latency measurements are now captured and displayed in analytics data.

Reviewer: ```suggestion
 Resolved an issue where Host and Latency fields (Total and Upstream) were not correctly reported for Tyk Gateways in MDCB data planes. The fix ensures accurate Host values and Latency measurements are now captured and displayed in the generated traffic logs.
```

>  </details>

Reviewer: ```suggestion
</details>
```

>  </li>

Reviewer: ```suggestion
</li>
```
